pollard_test.go

```diff
@@ -1339,3 +1340,148 @@ func compareNodeMap(mapA, mapB map[miniHash]*polNode) error {
+			numDels:         3,
+			delIndices:      []uint64{0, 1, 7},
+			rememberIndices: []uint64{0, 1, 2, 3, 4, 7, 8, 9},
+		},
+		{
+			name:            "delete four nodes",
+			numAdds:         10,
+			numDels:         4,
+			delIndices:      []uint64{0, 1, 2, 3},
+			rememberIndices: []uint64{0, 1, 2, 3, 4, 5, 8, 9},
+		},
+		{
+			name:            "delete five nodes",
+			numAdds:         10,
+			numDels:         5,
+			delIndices:      []uint64{0, 2, 4, 6, 8},
+			rememberIndices: []uint64{0, 1, 2, 4, 5, 6, 8, 9},
+		},
+		{
+			name:            "delete six nodes",
+			numAdds:         10,
+			numDels:         6,
+			delIndices:      []uint64{0, 1, 3, 5, 7, 9},
+			rememberIndices: []uint64{0, 1, 2, 3, 4, 5, 6, 7, 9},
+		},
+		{
+			name:            "delete seven nodes",
+			numAdds:         10,
+			numDels:         7,
+			delIndices:      []uint64{0, 1, 2, 4, 6, 8, 9},
+			rememberIndices: []uint64{0, 1, 2, 4, 5, 6, 8, 9},
+		},
+		{
+			name:            "delete eight nodes",
+			numAdds:         10,
+			numDels:         8,
+			delIndices:      []uint64{0, 1, 2, 3, 5, 6, 8, 9},
+			rememberIndices: []uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			name:            "delete nine nodes",
+			numAdds:         10,
+			numDels:         9,
+			delIndices:      []uint64{0, 1, 2, 3, 4, 6, 7, 8, 9},
+			rememberIndices: []uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		},
+		{
+			name:            "delete ten nodes",
+			numAdds:         10,
+			numDels:         10,
+			delIndices:      []uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+			rememberIndices: []uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Create a new accumulator with 10 leaves.
+			fmt.Println("\nTest case:", test.name)
+
+			adds := make([]Leaf, test.numAdds)
+
+			// Create the accumulator and add the leaves.
+			p := NewAccumulator(false)
+
+			for i := range adds {
+				if contains(test.rememberIndices, uint64(i)) {
+					adds[i] = Leaf{Hash: sha256.Sum256([]byte{uint8(i)}), Remember: true}
+				} else {
+					adds[i] = Leaf{Hash: sha256.Sum256([]byte{uint8(i)})}
+				}
+			}
+
+			err := p.Modify(adds, nil, Proof{})
+			if err != nil {
+				fmt.Println("Failed to add node:", err)
+			}
+
+			fmt.Printf("Merkle tree:\n\n%s\n", p.String())
+
+			// Delete the specified nodes.
+			dels := make([]Hash, test.numDels)
+			for i, idx := range test.delIndices {
+				dels[i] = adds[idx].Hash
+			}
+			proof, err := p.Prove(dels)
+			if err != nil {
+				t.Fatalf("Failed to generate proof: %v", err)
+			}
+			err = p.Modify(nil, dels, proof)
+			if err != nil {
+				t.Fatalf("Failed to delete nodes: %v", err)
+			}
+
+			fmt.Printf("Merkle tree after deletion of %d nodes:\n\n%s\n", test.numDels, p.String())
+
+			// Range through the target nodes (rememberIndices) and for the nodes
+			// which are not yet deleted, fetch their proof nodes and ensure that they exist
+			for _, i := range test.rememberIndices {
```

Comment:
This test isn't correct because you're looping through the positions of the remembers that happened after the Modify.

```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
00  01  02  03  04  05  06  07
```

For that tree, when you delete `00`, `01` moves up to `08`. The tree now looks like so:



```
14
|---------------\
12              13
|-------\       |-------\
08      09      10      11
|---\   |---\   |---\   |---\
        02  03  04  05  06  07
```

So if you're using the positions that were there before the deletion, the test isn't correct.

Reply:
So in the above example, do we include the position `08` in the nodes that we test?

In the current test, I am ignoring `01` completely since it has moved up and the `getNode` function for `01` returns `nil`.

If we have to include that, how do I get the node value? Is there a function defined?